packages/devtools_app/lib/src/memory/memory_events_pane.dart

```diff
@@ -0,0 +1,255 @@
+    _ghostTopLineSet.setScatterShapeSize(0);
+
+    // Dataset for VM GCs.
+    _gcVmDataSet = ScatterDataSet(eventsData.gcVm, 'VM GC');
+    _gcVmDataSet.setScatterShape(ScatterShape.CIRCLE);
+    _gcVmDataSet.setScatterShapeSize(6);
+    _gcVmDataSet.setScatterShapeHoleColor(ColorUtils.HOLO_BLUE_DARK);
+
+    // Dataset for user generated GC.
+    _gcUserDataSet = ScatterDataSet(eventsData.gcUser, 'User GC');
+    _gcUserDataSet.setScatterShape(ScatterShape.CIRCLE);
+    _gcUserDataSet.setScatterShapeHoleRadius(.6);
+    _gcUserDataSet.setScatterShapeHoleColor(ColorUtils.HOLO_BLUE_LIGHT);
+
+    // Dataset for user generated Snapshot.
+    _snapshotDataSet = ScatterDataSet(eventsData.snapshot, 'Snapshot');
+    _snapshotDataSet.setScatterShape(ScatterShape.TRIANGLE);
+    _snapshotDataSet.setScatterShapeHoleRadius(.6);
+    _snapshotDataSet.setScatterShapeHoleColor(ColorUtils.HOLO_GREEN_DARK);
+    _snapshotDataSet.setColor1(ColorUtils.HOLO_GREEN_DARK);
+
+    _allocationStartSet = ScatterDataSet(eventsData.monitorStart, 'Monitor Start');
+    _allocationStartSet.setScatterShape(ScatterShape.SQUARE);
+    _allocationStartSet.setScatterShapeSize(10);
+    _allocationStartSet.setColor1(ColorUtils.YELLOW);
+
+    _allocationContinueSet = ScatterDataSet(eventsData.monitorContinues, 'Monitor Continue');
+    _allocationContinueSet.setScatterShape(ScatterShape.SQUARE);
+    _allocationContinueSet.setScatterShapeSize(8);
+    _allocationContinueSet.setColor2(ColorUtils.YELLOW, 30);
+
+    _allocationResetSet = ScatterDataSet(eventsData.monitorReset, 'Monitor Reset');
+    _allocationResetSet.setScatterShape(ScatterShape.CIRCLE);
+    _allocationResetSet.setScatterShapeHoleRadius(.8);
+    _allocationResetSet.setScatterShapeHoleColor(ColorUtils.YELLOW);
+    _allocationResetSet.setColor2(ColorUtils.YELLOW, 110);
+
+    // Datset for automatic Snapshot.
+    _snapshotAutoDataSet = ScatterDataSet(
+      eventsData.snapshotAuto,
+      'Snapshot-Auto',
+    );
+    _snapshotAutoDataSet.setScatterShape(ScatterShape.TRIANGLE);
+    _snapshotAutoDataSet.setScatterShapeHoleRadius(.6);
+    _snapshotAutoDataSet.setScatterShapeHoleColor(ColorUtils.HOLO_RED_LIGHT);
+    _snapshotAutoDataSet.setColor1(ColorUtils.HOLO_RED_LIGHT);
+
+    // create a data object with the data sets
+    _controller.data = ScatterData.fromList([
+      _ghostTopLineSet,
+      _gcVmDataSet,
+      _allocationStartSet,
+      _allocationContinueSet,
+      _allocationResetSet,
+      _gcUserDataSet,
+      _snapshotDataSet,
+      _snapshotAutoDataSet,
+    ]);
+    _controller.data.setDrawValues(false);
+  }
+
+  /// Loads all heap samples (live data or offline).
+  void _processAndUpdate([bool reloadAllData = false]) {
+    setState(() {
+      // Display new events in the pane.
+      _updateEventPane();
+    });
+  }
+
+  /// Display any newly received events in the chart.
+  void _updateEventPane() {
+    setState(() {
+      _controller.data = ScatterData.fromList([
```

Comment:
Nit:
`final dataSets = [
 _gcVmDataSet,
        _allocationStartSet,
        _allocationContinueSet,
        _allocationResetSet,
        _gcUserDataSet,
        _snapshotDataSet,
        _snapshotAutoDataSet,
];
_controller.data = ScatterData.fromList(dataSets);
for (var dataSet in dataSets) {
  dataSet.notifyDataSetChanged();
}
```

Reply:
Rewrote all charting to handle list of datasets with enums code is simplified and more robust too.